src/oop/special_class_method/medium/q2/MainTest.java

```diff
@@ -0,0 +1,236 @@
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(setAttributeMethodName2, new Class<?>[]{boolean.class}), String.format(incorrectMethodDefinition, getAttributeMethodName2, className));
+        assertTrue(testClass.hasModifier(setAttributeMethodName2, new Class<?>[]{boolean.class}, "public"), String.format(incorrectModifierMessage, getAttributeMethodName2, className));
+        assertTrue(testClass.hasReturnType(setAttributeMethodName2, new Class<?>[]{boolean.class}, Void.TYPE), String.format(incorrectReturnType, getAttributeMethodName2, className));
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, boolean.class}
+        };
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName2, setMethodArguments, testInstance);
+        String incorrectSetterMessage =
+                "Your " + setAttributeMethodName2 + " method does not correctly update the value of " + attributeName2 + ".";
+        _assertEquals(updatedValue, testClass.getFieldValue(testInstance, attributeName2),
+                incorrectSetterMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setAttributeMethodName2, "method should not return any output"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void correctToStringMethod(double value1, boolean value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, boolean.class}
+        };
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod("toString", null), String.format(incorrectMethodDefinition, "toString", className));
+        assertTrue(testClass.hasModifier("toString", null, "public"), String.format(incorrectModifierMessage, "toString", className));
+        assertTrue(testClass.hasReturnType("toString", null, String.class), String.format(incorrectReturnType, "toString", className));
+        String incorrectToStringMessage = "Your toString method does not return the correct String.";
+        Object testInstance = testClass.createInstance(arguments);
+        String expectedOutput = "My brightness is " + value1 + " and isOn is " + value2;
+        Object toStringOutput = testClass.callMethod("toString", testInstance);
+        _assertEquals(expectedOutput, toStringOutput, incorrectToStringMessage);
```

Comment:
We should use the `_assertEquals()` method that accepts a delta so that if there is some rounding error with the computer, the student is not docked marks.

Reply:
This goes for all the double asserts in this file.

Reply:
Just make the delta 0.001 or something similar